src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -66,6 +74,14 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
+            // If there is an unhandled exception, we would like to avoid setting tempdata as 
+            // the end user is going to see an error page anyway and also it helps us in avoiding
+            // accessing resources like Session too late in the request lifecyle where SessionFeature might
+            // not be available.
+            if (!context.HttpContext.Response.HasStarted && context.Exception != null)
```

Comment:
Could we stash an instance of the context in the request rather than storing individual state pieces?

Reply:
👍 Instead of polluting with multiple items in HttpContext, having a single context of tempdata in HttpContext is good.

Reply:
You also need to check `ExceptionHandled` https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/ResourceExecutedContext.cs#L103

Reply:
I initially had it, but after discussing with @Eilon removed the check. The idea is that what would a developer do if they had handled the exception, like they might send a custom error page. Since this is still not a successful flow, trying to save tempdata even in this scenario didn't seem right/necessary.

Reply:
Yeah I don't think we need to check ExceptionHandled. Even if it was handled, would you really want the TempData to be maintained?

I could probably be convinced otherwise - I don't feel super strongly that one way is significantly better than the other, I'm just guessing it's the "more likely correct" answer.

Reply:
I don't feel too strongly about this other than to point out that usually the roles would be reversed - I'd be arguing for staying close to the intent, and you'd be pointing out that this is a breaking change 😆

Reply:
Yeah it's a weird kind of case. Do we assume the ship is going down, or is it worth rescuing... (Assume James Cameron does not exist in this universe.)